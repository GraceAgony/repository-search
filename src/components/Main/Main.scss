@import "./../../colors.scss";

.c-app {
  display: flex;
  flex-direction: column;
  max-width: 100vw;
  width: 900px;
  margin: 20px auto;
  align-items: center;

  &__error {
    border: $bg-red 2px solid;
    background-color: $bg-red-light;
    border-radius: 5px;
    padding: 1em;
    margin-top: 230px;
    color: $bg-red;
    max-width: 100%;
  }

  &__logo {
    max-width: 100%;
    width: 200px;
    display: block;
    margin: auto;
  }

  &__head {
    position: fixed;
    background-color: $bg-white;
    width: 100%;
    top: 0;
  }

  &__empty-search {
    margin-top: 230px;
  }
}
