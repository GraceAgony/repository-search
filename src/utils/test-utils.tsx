import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import { rootReducer } from "./../redux/store";
import { initialState as reducerInitialState } from "./../redux/modules/repos/reposReducer";

function render(
  ui: JSX.Element,
  {
    initialState = { repositories: reducerInitialState },
    store = createStore(rootReducer, initialState, applyMiddleware(thunk)),
    ...renderOptions
  }: { initialState?: any; store?: any } = {}
) {
  function Wrapper({ children }: any) {
    return <Provider store={store}>{children}</Provider>;
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}

export * from "@testing-library/react";
export { render };
