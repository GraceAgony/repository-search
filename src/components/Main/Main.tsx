import * as React from "react";
import { useSelector } from "react-redux";

import Search from "../Search";
import Paginator from "components/Paginator";

import octocat from "../../images/Octocat.png";
import {
  selectSearchString,
  selectWereReposLoaded,
} from "redux/modules/repos/reposSelectors";

import "./Main.scss";

interface MainProps {}

const Main: React.FunctionComponent<MainProps> = (props) => {
  const isSearchString = useSelector(selectSearchString);
  const isLoading = !useSelector(selectWereReposLoaded);

  return (
    <div className="c-app">
      <div className="c-app__head">
        <img src={octocat} alt="github logo" className="c-app__logo" />
        <Search></Search>
      </div>
      {isLoading || isSearchString ? (
        <Paginator></Paginator>
      ) : (
        <div className="c-app__empty-search">
          Enter the name of the repository you want to find
        </div>
      )}
    </div>
  );
};

export default Main;
