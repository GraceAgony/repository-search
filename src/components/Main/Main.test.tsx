import React from "react";
import { render, cleanup, screen } from "../../utils/test-utils";
import Main from "./Main";

afterEach(cleanup);

test("should write a snapshot", () => {
  const { container } = render(<Main />);

  expect(container).toMatchSnapshot();
});

test("render error message when error in redux store", () => {
  render(<Main />, {
    initialState: {
      repositories: {
        error: new Error("error msg"),
        loading: false,
        items: {},
        currentSearchString: "some search",
      },
    },
  });
  expect(screen.getByText("Error: error msg")).toBeVisible();
});

test("render Paginator when repositories loading", () => {
  const { container } = render(<Main />, {
    initialState: {
      repositories: {
        error: null,
        loading: true,
        items: {},
        currentSearchString: "some search",
      },
    },
  });
  expect(container.querySelector(".c-paginator__loading")).toBeVisible();
});
