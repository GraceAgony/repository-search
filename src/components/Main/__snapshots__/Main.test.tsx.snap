// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should write a snapshot 1`] = `
<div>
  <div
    class="c-app"
  >
    <div
      class="c-app__head"
    >
      <img
        alt="github logo"
        class="c-app__logo"
        src="Octocat.png"
      />
      <div
        class="c-search"
      >
        <input
          aria-label="search-input"
          class="c-search__input"
          placeholder="Search"
          type="search"
        />
        <button
          class="c-search__button"
        >
          <svg
            height="30"
            viewBox="0 0 172 172"
            width="30"
          >
            <g
              fill="none"
              font-family="none"
              font-size="none"
              font-weight="none"
              stroke-miterlimit="10"
              style="mix-blend-mode: normal;"
              text-anchor="none"
            >
              <path
                d="M0 172V0h172v172z"
              />
              <path
                d="M64.5 14.333c-27.621 0-50.167 22.546-50.167 50.167 0 27.621 22.546 50.167 50.167 50.167 12.527 0 23.973-4.673 32.782-12.318l3.051 3.051v9.267l43 43 14.334-14.334-43-43H105.4l-3.051-3.051c7.645-8.81 12.318-20.255 12.318-32.782 0-27.621-22.546-50.167-50.167-50.167zm0 14.334A35.725 35.725 0 01100.333 64.5 35.725 35.725 0 0164.5 100.333 35.725 35.725 0 0128.667 64.5 35.725 35.725 0 0164.5 28.667z"
                fill="#fff"
              />
            </g>
          </svg>
        </button>
      </div>
    </div>
    <div
      class="c-app__empty-search"
    >
      Enter the name of the repository you want to find
    </div>
  </div>
</div>
`;
